daml-lf/tests/BasicTests.daml

```diff
@@ -615,4 +615,81 @@ template NoMaintainer
     nonconsuming choice Noop: ()
       controller sig
       do
-        pure ()
\ No newline at end of file
+        pure ()
+
+template DuplicateKeyOperations
+  with
+    p : Party
+  where
+    signatory p
+
+    choice DuplicateBothTransient : ()
+      controller p
+      do cid1 <- create (WithKey p 42)
+         cid2 <- create (WithKey p 42)
+         archive cid1
+         archive cid2
+
+    choice DuplicateNoArchive : ()
+      controller p
+      do _ <- create (WithKey p 42)
+         _ <- create (WithKey p 42)
+         pure ()
+
+    choice DuplicateGlobalFetchByKey : ()
+      controller p
+      do _ <- fetchByKey @WithKey (p, 42)
+         cid1 <- create (WithKey p 42)
+         archive cid1
+
+    choice DuplicateGlobalFetchByKeyAfter : ()
+      controller p
+      do cid1 <- create (WithKey p 42)
+         _ <- fetchByKey @WithKey (p, 42)
```

Comment:
Many familiar with database transactions would expected fetchByKey return cid1 (no duplicate).

Reply:
It does. This does not fail in speedy, it is internally consistent. But it should fail in the ledger if there already exists another contract at the point of creation of `cid1`.

Reply:
Cool.  I see now this is a success criteria not a failure one.